worlds/sc2/rules.py

```diff
@@ -2394,15 +2538,56 @@ def __init__(self, world: Optional['SC2World']):
         self.basic_terran_units = get_basic_units(self.logic_level, SC2Race.TERRAN)
         self.basic_zerg_units = get_basic_units(self.logic_level, SC2Race.ZERG)
         self.basic_protoss_units = get_basic_units(self.logic_level, SC2Race.PROTOSS)
+        self.spear_of_adun_presence = SpearOfAdunPresence.default if world is None else world.options.spear_of_adun_presence.value
         self.spear_of_adun_autonomously_cast_presence = get_option_value(world, "spear_of_adun_autonomously_cast_ability_presence")
         self.enabled_campaigns = get_enabled_campaigns(world)
         self.mission_order = get_option_value(world, "mission_order")
         self.generic_upgrade_missions = get_option_value(world, "generic_upgrade_missions")
 
+
 def get_basic_units(logic_level: int, race: SC2Race) -> Set[str]:
     if logic_level == RequiredTactics.option_no_logic:
         return no_logic_basic_units[race]
     elif logic_level == RequiredTactics.option_advanced:
         return advanced_basic_units[race]
     else:
         return basic_units[race]
+
+
+def has_terran_units(player: int, target: int) -> Callable[['CollectionState'], bool]:
+    def _has_terran_units(state: CollectionState) -> bool:
+        return (
+            state.count_from_list_unique(item_groups.terran_units + item_groups.terran_buildings, player) >= target
+        )
+    return _has_terran_units
+
+
+def has_zerg_units(player: int, target: int) -> Callable[['CollectionState'], bool]:
+    def _has_zerg_units(state: CollectionState) -> bool:
+        return (
+            state.count_from_list_unique(item_groups.zerg_units + item_groups.zerg_buildings, player) >= target
+        )
+    return _has_zerg_units
+
+
+def has_protoss_units(player: int, target: int) -> Callable[['CollectionState'], bool]:
+    def _has_protoss_units(state: CollectionState) -> bool:
+        return (
+            state.count_from_list_unique(
+                item_groups.protoss_units
+                + item_groups.protoss_buildings
+                + [item_names.NEXUS_OVERCHARGE],
+                player
+            ) >= target
+        )
```

Comment:
`ValidInventory` doesn't understand `count_from_list_unique`

Reply:
I have already added `count_from_list_unique()` to ValidInventory in this PR. Though I suppose it wasn't added to that one test case that does fuzzing.